ext/opcache/zend_accelerator_util_funcs.c

```diff
@@ -156,6 +159,9 @@ static void zend_accel_function_hash_copy(HashTable *target, HashTable *source)
 			goto failure;
 		}
 		_zend_hash_append_ptr_ex(target, p->key, Z_PTR(p->val), 1);
+		if (UNEXPECTED(call_observers) && *ZSTR_VAL(p->key)) { // if not rtd key
```

Comment:
Maybe move the `UNEXPECTED(call_observers)` check into `zend_observer_function_declared_notify`, mark it as `zend_always_inline` and move the `for` loop inside it into a separate never inlined function? This way we get that  check in all cases without causing a function call in the common, negative case. The same could be done for the class declaration observer.

Reply:
In this specific case we want to check for call_observers before checking for the rtd key (which can easily be mispredicted).

Also, the *_notify functions access the observer llists, which are all not ZEND_API exported - so we would have to add these to the APIs, if we want to move these to the headers and make them inlineable.

Reply:
Ok. The content of `p->key` (hash) is accessed above, so the page of the first string char is likely in CPU cache, but of course I can't say if that's enough to not make the order matter.

> Also, the *_notify functions access the observer llists, which are all not ZEND_API exported

As mentioned I'd move the `for` over the observer list into it's own function and call that from the inlined function. With observers the additional function call probably doesn't matter, but it can be avoided when observers aren't enabled.

Reply:
(I'm talking about a mispredict, not the memory fetch)

Ah, I see, misread that. But then I think in general this does not matter too much.

Reply:
> (I'm talking about a mispredict, not the memory fetch)

As in branch prediction? I would've assumed that the `condition1 && UNEXPECTED(condition2)` would be considered unlikely overall (as both conditions need to be met). But I haven't looked at the actual generated code.

Reply:
My last comment didn't really make sense. `condition1` is a bad predictor for guessing if `condition2` or the statement after the `if` should be loaded. I thought maybe the compiler flips the conditions if they don't have side effects but it doesn't look like it . https://godbolt.org/z/5bWYWMMGY

Reply:
Yeah, it only means that "if condition1 is true, then condition2 is likely".

Reply:
I think the current way is the fastest you'll get.